HW01 Колмогорова Валерия

rerapony opened the pull request.

rerapony committed:
hw01

FirstTimeInForever commented on HW1/src/command.py:
```diff
@@ -0,0 +1,7 @@
+class Command:
+    """The Command interface"""
+
+    @staticmethod
+    def execute(*args):
```
Почему static?

rerapony replied:
fixed

FirstTimeInForever commented on HW1/src/parser.py:
```diff
@@ -0,0 +1,109 @@
+    """
+    def __init__(self):
+        self.globals = defaultdict()
+        self.command = defaultdict()
+        self.command["cat"] = Cat.execute
+        self.command["echo"] = Echo.execute
+        self.command["exit"] = Exit.execute
+        self.command["pwd"] = Pwd.execute
+        self.command["wc"] = Wc.execute
```
Нужно вынести в [абстрактную фабрику](https://refactoring.guru/design-patterns/abstract-factory).

rerapony replied:
done

FirstTimeInForever reviewed (changes requested):
Документация:
* Нет описания приложения в целом и применённых архитектурных решений
* Не указано как запускать
* Нет описания поддерживаемых команд и их аргументов

Архитектура - с ней есть такие проблемы:
* Намертво прибиты `sys.stdin`/`sys.stdout`
* Нет возможности легко менять набор доступных команд
* `exit` убивает всё приложение
Чтобы это исправить нужно поделить весь код на сущности:
* `CommandFactory` - [абстрактная фабрика](https://refactoring.guru/design-patterns/abstract-factory) команд. Только фабрика должна знать о существующих командах. На всех нижних уровнях приложения команда - это просто список токенов.
* `Parser` - получает на вход строку и на выходе отдаёт набор осмысленных токенов.
* `Shell` - содержит input loop. Параметризируется инстансом фабрики и, может быть, инстансом парсера. Забирает у парсера осмысленные токены и отдаёт их фабрике. Фабрика отдаёт сконструированную команду или ошибку.

Чтобы абстрагироваться от конкретных IO потоков нужно параметризировать `Shell` и `Command` нужными потоками. Для `exit` можно бросать `ExitException` и ловить его в input loop. После этого нужно спокойно передать управление вызывающему коду.

Функционал:
* Не поддерживаются внешние команды:
```
> ls
Command ls is not found

> git status
Command ['git', 'status'] is not found
```
* Не работают пути вида (и выводятся аргументы):
```
> cat ../README.md
['../README.md']
Incorrect arguments: ['../README.md']
```
* Не работает пайп:
```
> cat main.py | wc
main.py
Incorrect arguments
: main.py
```
* Неправильно работают подстановки:
```
> x=1
> echo $x
1
> echo "$x"
1
> echo '$x'
1
> echo "'$x'"
1
```
```
> x=ex
> y=it
> $x$y
Command $x$y is not found
```
* Нет обработки ошибок:
```
> pwd asd
Traceback (most recent call last):
  File "/Users/hans/stuff/trash/software-design-2021/HW1/main.py", line 17, in <module>
    main()
  File "/Users/hans/stuff/trash/software-design-2021/HW1/main.py", line 13, in main
    parser.parse(cmd)
  File "/Users/hans/stuff/trash/software-design-2021/HW1/src/parser.py", line 96, in parse
    result = command(args)
TypeError: execute() takes 0 positional arguments but 1 was given
```
```
> echo some | cat
Traceback (most recent call last):
  File "/Users/hans/stuff/trash/software-design-2021/HW1/main.py", line 17, in <module>
    main()
  File "/Users/hans/stuff/trash/software-design-2021/HW1/main.py", line 13, in main
    parser.parse(cmd)
  File "/Users/hans/stuff/trash/software-design-2021/HW1/src/parser.py", line 87, in parse
    command = self.parse_command(arg)
  File "/Users/hans/stuff/trash/software-design-2021/HW1/src/parser.py", line 107, in parse_command
    if command in self.command.keys():
TypeError: unhashable type: 'list'
```

Плюс:
* Нет тестов
* Нет CI
* Нет линтера/энфорсинга кодстайла

Пожалуйста переделай архитектуру, исправь функционал и добавь недостающие части задания.

rerapony committed:
fixed architecture

rerapony committed:
dirs names refactoring

rerapony committed:
Create main.yml

rerapony committed:
added pylint check

rerapony committed:
added tests, fixed pipe, fixed $ substitution, fixed architecture

rerapony commented:
> * Намертво прибиты `sys.stdin`/`sys.stdout`

fixed 



> * `exit` убивает всё приложение

а разве так и не должно быть? exit в баше убивает терминал же 



> * Чтобы это исправить нужно поделить весь код на сущности:

сделано, переписала всё нафиг



> * Не работает пайп:

fixed

> * Неправильно работают подстановки:

fixed

> > pwd asd

fixed 



> * Нет тестов

fixed



> * Нет CI

fixed (pylint)

rerapony committed:
fixed pylint

rerapony committed:
.gitignore

rerapony committed:
fixed readme

FirstTimeInForever reviewed (commented):
Не запускается:
```
python3 ./run_cli.py
Traceback (most recent call last):
  File "/Users/hans/stuff/trash/software-design-2021/cli/./run_cli.py", line 3, in <module>
    from cli.src.factory.command_factory import CommandFactory
ModuleNotFoundError: No module named 'cli'
```

Архитектура:
Абстрактная фабрика работает не так. Идейно нужно что-то вроде такого:
```
class MyFactory: MyFactoryInterface {
  fun createCommand(name: String, arguments: List<String>): Command {
    return when (name) {
      "pwd" -> PwdCommand(arguments)
      "exit" -> ExitCommand(arguments)
      ...
      else -> ExternalCommand(arguments)
    }
  }
}

class Shell(commandFactory: MyFactoryInterface) { ... }
```
Тогда можно спокойно заменять имплементацию фабрики чтобы поменять множество доступных команд.


Функционал:
* Не работают подстановки с несколькими переменными:
```
x=1
echo $x$x
Traceback (most recent call last):
  File "/Users/hans/stuff/trash/software-design-2021/cli/run_cli.py", line 18, in <module>
    run_cli(sys.stdin, sys.stdout)
  File "/Users/hans/stuff/trash/software-design-2021/cli/run_cli.py", line 14, in run_cli
    shell.run_shell()
  File "/Users/hans/stuff/trash/software-design-2021/cli/src/shell.py", line 24, in run_shell
    tokens_list = self.parser.parse(line)
  File "/Users/hans/stuff/trash/software-design-2021/cli/src/parser.py", line 25, in parse
    line = self.substitute(line)
  File "/Users/hans/stuff/trash/software-design-2021/cli/src/parser.py", line 58, in substitute
    raise Exception("Incorrect statement")
Exception: Incorrect statement
```
* Нет обработки ошибок неизвестных команд:
```
asd
Traceback (most recent call last):
  File "/Users/hans/stuff/trash/software-design-2021/cli/src/shell.py", line 35, in run_shell
    command = getattr(self.factory, command_name)()
AttributeError: 'CommandFactory' object has no attribute 'asd_cmd'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/hans/stuff/trash/software-design-2021/cli/run_cli.py", line 18, in <module>
    run_cli(sys.stdin, sys.stdout)
  File "/Users/hans/stuff/trash/software-design-2021/cli/run_cli.py", line 14, in run_cli
    shell.run_shell()
  File "/Users/hans/stuff/trash/software-design-2021/cli/src/shell.py", line 37, in run_shell
    raise Exception("Bad command: {}".format(command_name)) from exc
Exception: Bad command: asd_cmd
```
* Не работают длинные пайплайны:
```
cat run_cli.py | wc
19 38 413
cat run_cli.py | wc | wc
Traceback (most recent call last):
  File "/Users/hans/stuff/trash/software-design-2021/cli/run_cli.py", line 18, in <module>
    run_cli(sys.stdin, sys.stdout)
  File "/Users/hans/stuff/trash/software-design-2021/cli/run_cli.py", line 14, in run_cli
    shell.run_shell()
  File "/Users/hans/stuff/trash/software-design-2021/cli/src/shell.py", line 39, in run_shell
    result = command.execute(result, count, *args)
  File "/Users/hans/stuff/trash/software-design-2021/cli/src/command/wc.py", line 30, in execute
    raise ValueError("Invalid path: {}".format(path))
ValueError: Invalid path: 
```
* Не вызываются внешние команды:
```
git status
Traceback (most recent call last):
  File "/Users/hans/stuff/trash/software-design-2021/cli/src/shell.py", line 35, in run_shell
    command = getattr(self.factory, command_name)()
AttributeError: 'CommandFactory' object has no attribute 'git_cmd'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/hans/stuff/trash/software-design-2021/cli/run_cli.py", line 18, in <module>
    run_cli(sys.stdin, sys.stdout)
  File "/Users/hans/stuff/trash/software-design-2021/cli/run_cli.py", line 14, in run_cli
    shell.run_shell()
  File "/Users/hans/stuff/trash/software-design-2021/cli/src/shell.py", line 37, in run_shell
    raise Exception("Bad command: {}".format(command_name)) from exc
Exception: Bad command: git_cmd
```
* Нет сообщения об ошибке из-за непарных кавычек:
```
echo "

echo '
Traceback (most recent call last):
  File "/Users/hans/stuff/trash/software-design-2021/cli/run_cli.py", line 18, in <module>
    run_cli(sys.stdin, sys.stdout)
  File "/Users/hans/stuff/trash/software-design-2021/cli/run_cli.py", line 14, in run_cli
    shell.run_shell()
  File "/Users/hans/stuff/trash/software-design-2021/cli/src/shell.py", line 24, in run_shell
    tokens_list = self.parser.parse(line)
  File "/Users/hans/stuff/trash/software-design-2021/cli/src/parser.py", line 31, in parse
    raise Exception("Incorrect command: odd number of single quotes")
Exception: Incorrect command: odd number of single quotes
```

* Хотелось бы больше тестов

`3/7`

FirstTimeInForever commented:
> а разве так и не должно быть? exit в баше убивает терминал же

Я немного не пояснил: убийство всего процесса в этой команде не передаст управление вызывающему коду. После получения сигнала о необходимости выхода, `Shell` мог бы выполнять какую-нибудь полезную работу, например, закрывать захваченные ресурсы. Ну и просто, если бы этот шелл был бы билиотекой/модулем, который использовался бы кем-то, то получать `exit()` было бы странно.

Решить это можно, например:
* возвращать вместе с результатом команды `bool` - продолжать выполнение или завершать
* тоже самое, но устанавливать где-то флаг (только не глобальный)
* бросать исключение `ExitException` и ловить его в вызывающем коде
